### PR TITLE
Domain Metadata Exporter

### PR DESCRIPTION
**Pull request by kasenvr**
image

What this essentially does is make available the metadata to the public. Before it would only report metadata to its connected Metaverse API. It will still do that, but will also export to JSON on request.

What might this help with?

- Security whitelists for a domain https://github.com/kasenvr/project-athena/issues/368 so that people can use their whitelist and also "Accept" or "Deny" adding the domain's approved URLs to your whitelist while you are visiting.
- Shared ban lists, approved user lists, etc. Anything that you can think of wanting to share that you set with your domain or want to broadcast from your domain, you can do it with this. There are currently not many fields to edit your metadata with at this point in time, but boxes for that can be added later. This sets the stage for this.
- Detaching from a Metaverse API or having services in parallel to one without having to have the Metaverse API be programmed to handle/route such things.

**This PR is based on PR https://github.com/kasenvr/project-athena/pull/280 and therefore that should be merged first before addressing this one.**

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-560e4d7_0c53701b.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-560e4d7.tar_0c53701b.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-560e4d7.tar_0c53701b.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-560e4d7_0c53701b.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-560e4d7_0c53701b.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-a8b0d2e_f712fc77.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-a8b0d2e.tar_f712fc77.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-a8b0d2e.tar_f712fc77.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-a8b0d2e_f712fc77.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-a8b0d2e_f712fc77.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-3ea6a30.tar_f1d731da.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-3ea6a30_f1d731da.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-3ea6a30.tar_f1d731da.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-3ea6a30_f1d731da.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-3ea6a30_f1d731da.dmg

File: domain-server/resources/metadata_exporter/index.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<html>
+	<head>
+		<title>Vircadia Metadata exporter</title>
+	</head>
+
+	<body>
+		<h1>Vircadia Metadata exporter</h1>
```
Comment (ctrlaltdavid) [inline domain-server/resources/metadata_exporter/index.html:7]:
```suggestion
		<h1>Vircadia Metadata Exporter</h1>
```

File: domain-server/src/DomainMetadata.h [hunk 2]
```diff
@@ -42,16 +43,17 @@ Q_OBJECT
...
     DomainMetadata(QObject* domainServer);
-    DomainMetadata() = delete;
-
+    ~DomainMetadata() = default;
+    
     // Get cached metadata
     QJsonObject get();
     QJsonObject get(const QString& group);
 
+    bool handleHTTPRequest(HTTPConnection* connection, const QUrl& url, bool skipSubHandler = false);
```
Comment (ctrlaltdavid) [inline domain-server/src/DomainMetadata.h:52]:
```suggestion
    bool handleHTTPRequest(HTTPConnection* connection, const QUrl& url, bool skipSubHandler = false) override;
```

File: domain-server/src/DomainMetadata.h [hunk 3]
```diff
@@ -42,16 +43,17 @@ Q_OBJECT
     };
 
     DomainMetadata(QObject* domainServer);
-    DomainMetadata() = delete;
-
+    ~DomainMetadata() = default;
+
```
Comment (ctrlaltdavid) [inline domain-server/src/DomainMetadata.h:47]:
```suggestion

```

File: domain-server/src/DomainServer.cpp [hunk 4]
```diff
@@ -3060,6 +3071,26 @@ void DomainServer::initializeExporter()
...
 
+void DomainServer::initializeMetadataExporter() {
+    static const QString ENABLE_EXPORTER = "metaverse.enable_metadata_exporter";
+    static const QString EXPORTER_PORT = "metaverse.metadata_exporter_port";
+
+    bool isMetadataExporterEnabled = _settingsManager.valueOrDefaultValueForKeyPath(ENABLE_EXPORTER).toBool();
+    int metadataExporterPort = _settingsManager.valueOrDefaultValueForKeyPath(EXPORTER_PORT).toInt();
+
+    if (metadataExporterPort < MIN_PORT || metadataExporterPort > MAX_PORT) {
+        qCWarning(domain_server) << "Metadata exporter port " << metadataExporterPort << " is out of range.";
```
Comment (ctrlaltdavid) [inline domain-server/src/DomainServer.cpp:3082]:
```suggestion
        qCWarning(domain_server) << "Metadata exporter port" << metadataExporterPort << "is out of range.";
```
Spaces are automatically included between items.

File: domain-server/src/DomainServer.cpp [hunk 5]
```diff
@@ -3060,6 +3071,26 @@ void DomainServer::initializeExporter()
...
+        isMetadataExporterEnabled = false;
+    }
+
+    qCDebug(domain_server) << "Setting up Metadata exporter.";
+
+    if (isMetadataExporterEnabled && !_httpMetadataExporterManager) {
+        qCInfo(domain_server) << "Starting Metadata exporter on port " << metadataExporterPort;
```
Comment (ctrlaltdavid) [inline domain-server/src/DomainServer.cpp:3089]:
```suggestion
        qCInfo(domain_server) << "Starting Metadata exporter on port" << metadataExporterPort;
```

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-4c900a9.tar_1c545dff.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-4c900a9_1c545dff.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-4c900a9.tar_1c545dff.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-4c900a9_1c545dff.exe

File: domain-server/src/DomainServer.cpp [hunk 6]
```diff
@@ -3060,6 +3071,26 @@ void DomainServer::initializeExporter()
...
+    if (metadataExporterPort < MIN_PORT || metadataExporterPort > MAX_PORT) {
+        qCWarning(domain_server) << "Metadata exporter port" << metadataExporterPort << "is out of range.";
+        isMetadataExporterEnabled = false;
+    }
+
+    qCDebug(domain_server) << "Setting up Metadata exporter.";
+
+    if (isMetadataExporterEnabled && !_httpMetadataExporterManager) {
+        qCInfo(domain_server) << "Starting Metadata exporter on port" << metadataExporterPort;
+        _httpMetadataExporterManager = new HTTPManager(QHostAddress::Any, (quint16)metadataExporterPort, QString("%1/resources/metadata_exporter/").arg(QCoreApplication::applicationDirPath()), _metadata);
```
Comment (ctrlaltdavid) [inline domain-server/src/DomainServer.cpp:3090]: Line too long.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-135bc25.tar_eef1e3b9.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-135bc25.tar_eef1e3b9.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-135bc25_eef1e3b9.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-135bc25_eef1e3b9.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-135bc25_eef1e3b9.exe

**Comment by daleglass** [comment]
Ok, looking into how it runs, I get this:

* `description: null` -- that seems bad. A description is a string, and probably should just be an empty one by default. Nulls need to be dealt with specifically.
* `hosts: []` -- What are those?
* `maturity: null` -- same issue as above. Should have some reasonable default. Either an actual value, or perhaps something like "Not configured", if distinguishing domains where the owner hasn't filled the data in is important.
* `user_hostnames: * : 1` -- I don't get this one.

Other issues:

The structure shouldn't have an array of hashes, it should have clear naming. Otherwise, code has to deal with indexes like [1], rather than meaningful names like ["statistics"]

Domain metadata probably should have a "name" field. Then "name" is the general short name of the place, and "description" might be a paragraph or two worth.

Other useful fields could be: an URL, a list of rules, an owner's name, administrative contact info, an URL to an icon or logo, and list of URLs to screenshots.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-337291e_c70a0414.dmg

**Comment by kasenvr** [comment]
image

**Comment by kasenvr** [comment]
I still have not addressed user_hostnames, I'll have to look into that, those are raw values that are usually sent to the metaverse API.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-45abeba_b7e43625.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-45abeba.tar_b7e43625.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-45abeba_b7e43625.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-45abeba.tar_b7e43625.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR465/Vircadia-Alpha-PR465-45abeba_b7e43625.exe

**Comment by daleglass** [comment]
Discussed in dev meeting: it'd be nice to have testing and feedback from people who can take advantage of this functionality.

**Review by daleglass — APPROVED** [review]

**Comment by kasenvr** [comment]
I've tested it during development, Dale's run an earlier version of it, what I've added shouldn't really affect much other than... either it works or it doesn't. :D

**Comment by Misterblue** [comment]
In implementing this, I notice that meta.imags is an array of image names when the existing implementation has three named images: "hero", "tiny", and "thumbnail". So, 'meta.images' needs to changed from an array to an object formatted as:
{
    "tiny": "imageURL",
    "hero": "imageURL",
    "thumbnail": "imageURL"
}
or similar.

**Comment by Misterblue** [comment]
Also there is a lot of additional data that is not included in the sample. LIke domainId, place_name, public_key. Is that all considered 'meta'?

Also, there is currently no contact_info but there is an account associated with the domain -- the account that connected itself to the domain on startup. Is that the same thing?

**Comment by kasenvr** [comment]
I'll have to perform further updates to get that working. :)

**Pull request merged**